pallets/rc-migrator/src/lib.rs

```diff
@@ -0,0 +1,2581 @@
+
+			let current_stage = RcMigrationStage::<T>::get();
+			ensure!(
+				matches!(current_stage, MigrationStage::Scheduled { .. }),
+				Error::<T>::InvalidStageTransition
+			);
+
+			Self::transition(MigrationStage::Pending);
+
+			Self::deposit_event(Event::MigrationCancelled);
+
+			Ok(Pays::No.into())
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+		where
+		crate::BalanceOf<T>:
+			From<<<T as polkadot_runtime_common::slots::Config>::Currency as frame_support::traits::Currency<sp_runtime::AccountId32>>::Balance>,
+		crate::BalanceOf<T>:
+			From<<<<T as polkadot_runtime_common::crowdloan::Config>::Auctioneer as polkadot_runtime_common::traits::Auctioneer<<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number>>::Currency as frame_support::traits::Currency<sp_runtime::AccountId32>>::Balance>,
+		<<T as pallet_treasury::Config>::BlockNumberProvider as BlockNumberProvider>::BlockNumber: Into<u32>
+	{
+		fn integrity_test() {
+			let (ah_ump_priority_blocks, _) = T::AhUmpQueuePriorityPattern::get();
+			assert!(!ah_ump_priority_blocks.is_zero(), "the `ah_ump_priority_blocks` should be non-zero");
+		}
+
+		fn on_finalize(now: BlockNumberFor<T>) {
+			if Self::is_ongoing() {
+				Self::force_ah_ump_queue_priority(now);
+			}
+		}
+
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			let mut weight_counter = WeightMeter::with_limit(T::MaxRcWeight::get());
+
+			let stage = RcMigrationStage::<T>::get();
+			weight_counter.consume(T::DbWeight::get().reads(1));
+
+			if stage.is_ongoing() {
+				// account the weight of `on_finalize` for the `force_ah_ump_queue_priority` job.
+				weight_counter.consume(T::RcWeightInfo::force_ah_ump_queue_priority());
+			}
+
+			if Self::has_excess_unconfirmed_dmp(&stage) {
+				log::info!(
+					target: LOG_TARGET,
+					"Excess unconfirmed XCM messages, skipping the data extraction for this block."
+				);
+				return weight_counter.consumed();
+			}
+
+			match stage {
+				MigrationStage::Pending | MigrationStage::MigrationPaused { .. } => {
+					return weight_counter.consumed();
+				},
+				MigrationStage::Scheduled { start } => {
+					// Two sessions before the migration starts we pause staking election
+					let staking_pause_time = start.saturating_sub((T::SessionDuration::get().saturating_mul(2) as u32).into());
+
+					if now == staking_pause_time {
+						// stop any further staking elections
+						pallet_staking::ForceEra::<T>::put(pallet_staking::Forcing::ForceNone);
+						Self::deposit_event(Event::StakingElectionsPaused);
+					}
+
+					if now >= start {
+						weight_counter.consume(T::DbWeight::get().reads(2));
+
+						match Self::send_xcm(types::AhMigratorCall::<T>::StartMigration) {
+							Ok(_) => {
+								Self::transition(MigrationStage::WaitingForAh);
+							},
+							Err(_) => {
+								defensive!(
+									"Failed to send StartMigration message to AH, \
```

Comment:
we recover from this by either `force_set_stage` or `send_xcm_message`?

Reply:
It retries indefinitely so it would need manual intervention with `force_set_stage`, yes.